File: Array/217. Contains Duplicate/ContainsDuplicate.cpp
//
//  ContainsDuplicate.cpp
//  LeetCode-main
//
//  Created by jackma on 2022/2/16.
//  Copyright © 2022 JackMa. All rights reserved.
//

#include <unordered_set>

#include "ContainsDuplicate.hpp"

// my solution
bool Solution::containsDuplicate(vector<int> &nums) {
    unordered_set<int> set = unordered_set<int>();
    bool result = false;
    
    for (int num : nums) {
        // can't find num
        if (set.find(num) == set.end()) {
            set.insert(num);
        } else {
            result = true;
            break;
        }
        
    }
    return result;
}

// other solution
bool otherContainsDuplicate(vector<int> &nums) {
    std::sort(nums.begin(), nums.end());
    // Searches the range [first, last) for two consecutive equal elements.
    return std::adjacent_find(nums.begin(), nums.end()) != nums.end();
}

void testContainsDuplicate(void) {
    Solution solution = Solution();
    vector<int> nums = {1,2,3,4};
    cout << solution.containsDuplicate(nums) << endl;
}

File: Array/217. Contains Duplicate/ContainsDuplicate .swift
//
//  ContainsDuplicate .swift
//  LeetCode-Swift
//
//  Created by jackma on 2023/3/12.
//  Copyright © 2023 JackMa. All rights reserved.
//

import Foundation

/**
 https://leetcode.com/problems/contains-duplicate
 
 Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
 */

extension Solution {
    func ContainsDuplicate(_ nums: [Int]) -> Bool {
        var numsMap = [Int:Int]()
        
        for num in nums {
            if numsMap[num] == nil {
                numsMap[num] = 1
            } else {
                return true
            }
        }
        
        return false
    }
}
